### PR TITLE
Dev v2.9.0

### PR DESCRIPTION
**Pull request by rdilley-usgs**

### CODE REVIEW

**Comment by rdilley-usgs** [comment]
Don't go to the "Files changed" tab, it may take a while.....

File: README [hunk 1]
```diff
@@ -4,7 +4,23 @@ Author: David V. Hill, SGT
 License: NASA Open Source Agreement 1.3
 USGS Designation: EROS Science Processing Architecture (ESPA)
```
Comment (dhill-usgs) [inline README]: consider moving this to markdown...  you can use the readme in espa-web to get a jumpstart as much of it is the same for the history of changes.

**Comment by dhill-usgs** [comment]
I'd advise opening all your files in spyder and removing all the unused imports.

**Comment by dhill-usgs** [comment]
There's a bunch of unused variables too.  spyder will show them to you.

**Comment by dhill-usgs** [comment]
To help clean up some of your code, it's unnecessary to declare class member variables 'and' define them in init.  Consider defining them only in the __init__ method of your classes rather than in both places.

Example is your processor.PlotProcessor(ProductProcessor):

**Comment by dhill-usgs** [comment]
def get_ymds_from_filename(self, filename) is duplicating functionality in sensor.py

**Comment by dhill-usgs** [comment]
In the processor.py code, it would be helpful if your top level class defined all the expected methods + signatures and returned NotImplementedError, forcing the subclasses to override.  This would not only show clearly how to define a new processor (as a template) but would also make your code break in the event of a typo in a subclass (a good thing).

**Comment by dhill-usgs** [comment]
undefined variable name 'destination_product_file', line 519 in distribution.py.

**Comment by dhill-usgs** [comment]
sensor.py is defined in two places in the codebase... espa_common and in the processing/ dir

**Comment by dhill-usgs** [comment]
initialization.py... combine all the create_local_*_directory methods and just pass in the variable for the directory type that needs the be made.

**Comment by dhill-usgs** [comment]
i think you're already in the process of doing this, but try to work toward creating a consistent code structure and naming scheme for the modules and methods/functions, and then implementing the required functionality inside of that code framework.  Try to create 'strong' identities for not only the modules, but also the functions.  Anything that should not be called from outside of a module could be named with the double underscore prepended.

**Comment by dhill-usgs** [comment]
unit tests will help show how your code is intended to be used... (of course)

**Comment by dhill-usgs** [comment]
warp.py (and other places)

Consider not defining functions if the code isn't needed in more than one place.  Example: build_argument_parser appears to only be used in the __main__.  

Results in an unnecessary context switch mentally and in silicon.

**Comment by dhill-usgs** [comment]
warp.py is defining setting and importing settings.  Consider moving all this to your settings.py and naming the variables to make is clear what they're for (carve off a section of settings with comments, etc)

# These contain valid warping options
valid_resample_methods = ['near', 'bilinear', 'cubic', 'cubicspline',
                          'lanczos']
valid_pixel_size_units = ['meters', 'dd']
valid_image_extents_units = ['meters', 'dd']
valid_projections = ['sinu', 'aea', 'utm', 'ps', 'lonlat']
valid_ns = ['north', 'south']
# First entry in the datums is used as the default, it should always be set to
# WGS84
valid_datums = [settings.WGS84, settings.NAD27, settings.NAD83]

**Comment by dhill-usgs** [comment]
warp.py:1115  undefined variable (parms)

``` warp_espa_data(options, parms['scene'])```

**Comment by dhill-usgs** [comment]
statistics.py... the naming of get_statistics and generate_statistics and their functionality seems backwards to me (coming in from a naive point of view).  Would 'write_statistics' be a better description for what's happening in ```generate_statistics``` since ```get_statistics``` is actually generating them?

**Comment by dhill-usgs** [comment]
processor.py... redefinition of items that should be held only in your settings.py (and then, only once)
```
_WGS84 = 'WGS84'
_NAD27 = 'NAD27'
_NAD83 = 'NAD83'
```

**Comment by dhill-usgs** [comment]
Consistent language and use will help show the code design and ease maintenance

processor.py:
971 ```sr_command_line``` is building and returning a string (command line)
1228 ```build_science_products``` is driving the creation of science products

whereas warp.py:
101 ```build_sinu_proj4_string``` is building and returning a string

and ondemand_mapper.py:
107 ```process``` is driving creation

**Comment by dhill-usgs** [comment]
metadata.py

```get_landsat_metadata``` isn't actually getting the metadata, but the filename

**Comment by dhill-usgs** [comment]
espa_exception.py

Instead of generating error codes that mean something, I believe the Python thing to do would be to define specific named exceptions instead. e.g.  ```cfmask = 14``` becomes ```CFMaskException``` and so on.  This would be less error prone.

This is detailed in chapter 6 of Python in a Nutshell.... i can lend if necessary...

**Comment by dhill-usgs** [comment]
utilities.py

I show three copies of utilities.py in the codebase.  espa_common, processing and processing/test-orders.  Some of the methods contained in this module aren't used anywhere in the codebase.
* ```checksum_local_file```
* ```gzip_files```
* ```tar_files```
* ```create_link``` (used only in initialization.py)

**Comment by dhill-usgs** [comment]
settings.py

Remove variables that aren't used in the processing tier, such as ```SOAP_CACHE_LOCATION``` and ```SOAP_CLIENT_TIMEOUT```

**Comment by dhill-usgs** [comment]
readme... please try and capture all the pertinent environment variables to help us with setting up & adjusting code in your absence.

**Comment by dhill-usgs** [comment]
logger_factory.py

I think if you were to redefine ```class EspaLogging(object):``` to ```class EspaLogger(object):```, you would then be able to clean up all your underlying method names by changing things like ```def check_logger_configured(cls, logger_name):``` to ```def is_configured(cls, name)``` and ```def delete_logger_file(cls, logger_name):``` to ```def delete_file(cls, name)```.   

Just a suggestion.

**Comment by dhill-usgs** [comment]
That's probably enough for this review I think...  if you have questions let me know.

**Pull request merged**